wp_api/src/authenticator.rs

```diff
@@ -0,0 +1,310 @@
+        &self,
+        request: &WpNetworkRequest,
+        previous_response: &WpNetworkResponse,
+    ) -> Option<HeaderMap> {
+        if self.should_authenticate(&request.url.0, Some(previous_response.status_code)) {
+            self.reset();
+            return self.authentication_headers().await;
+        }
+
+        None
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct NilAuthenticator {}
+
+#[async_trait::async_trait]
+impl Authenticator for NilAuthenticator {
+    fn should_authenticate(&self, _request_url: &str, _response_status_code: Option<u16>) -> bool {
+        false
+    }
+
+    fn host(&self) -> &str {
+        ""
+    }
+
+    async fn authentication_headers(&self) -> Option<HeaderMap> {
+        None
+    }
+
+    fn reset(&self) {
+        // Do nothing.
+    }
+}
+
+#[derive(Debug)]
+pub struct ApplicationPasswordAuthenticator {
+    host: String,
+    token: String,
+}
+
+impl ApplicationPasswordAuthenticator {
+    pub fn new(host: String, token: String) -> Self {
+        Self { host, token }
+    }
+
+    pub fn with_application_password(host: String, username: String, password: String) -> Self {
+        use base64::prelude::*;
+        let token = BASE64_STANDARD.encode(format!("{}:{}", username, password));
+        Self::new(host, token)
+    }
+}
+
+#[async_trait::async_trait]
+impl Authenticator for ApplicationPasswordAuthenticator {
+    fn host(&self) -> &str {
+        self.host.as_str()
+    }
+
+    async fn authentication_headers(&self) -> Option<HeaderMap> {
+        let mut headers = HeaderMap::new();
+        headers.append(
+            http::header::AUTHORIZATION,
+            format!("Basic {}", self.token)
+                .try_into()
+                .expect("token is always a valid header value"),
+        );
+
+        Some(headers)
+    }
+
+    fn reset(&self) {
+        // Do nothing.
+    }
+
+    fn should_authenticate(&self, request_url: &str, response_status_code: Option<u16>) -> bool {
+        let is_same_host = Url::parse(request_url)
+            .ok()
+            .map(|f| f.host_str() == Some(self.host()));
+        if is_same_host != Some(true) {
+            return false;
+        }
+
+        // Authenticated has already been sent. No need to make a second attempt.
+        response_status_code.is_none()
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct CookieAuthenticator {
+    api_base_url: ApiBaseUrl,
+    credentials: WpLoginCredentials,
+    request_executor: std::sync::Arc<dyn RequestExecutor>,
+    nonce: RwLock<Option<String>>,
+}
+
+impl CookieAuthenticator {
+    pub(crate) fn new(
+        api_base_url: ApiBaseUrl,
+        credentials: WpLoginCredentials,
+        request_executor: std::sync::Arc<dyn RequestExecutor>,
+    ) -> Self {
+        Self {
+            api_base_url,
+            credentials,
+            request_executor,
+            nonce: None.into(),
+        }
+    }
+
+    async fn get_rest_nonce(&self) -> Option<String> {
```

Comment:
In my opinion, this function signature is misleading. At a glance, it seems to be returning a `nonce` token, but in fact it also mutates the state. At the very least we should name it such that it communicates the mutation.

This issue is mostly due to the interior mutability pattern we are using, so I'd suggest skipping this comment for now as I'd like to discuss the usage of `RwLock` in a separate thread first. We can come back to it afterwards if it's still applicable.

Reply:
I have renamed this function to `fetch_rest_nonce`. Let me know if you have any suggestions regarding the name.

Reply:
I like `fetch_rest_nonce`, but it looks like it's still mutating the hold `nonce` value inside. I'd strongly prefer if a fetch function only returns the value and the update happens somewhere else. If that's not a good option, then I was thinking we should go with something like `fetch_and_update_rest_nonce` instead.

Having said all that, I am looking at the implementation and it looks like the `nonce` value we hold is meant to be only accessed through the `fetch_rest_nonce` 🤔 

I have some concerns, but I don't have a clear suggestion at the moment, so let's leave this as is for this PR. I'll think about it again when/if I get to work on this particular code - or if we have any issues with it.